### ISSUE TITLE
Wikibase Bot: combining data types?

### ISSUE THREAD

**Issue by anchardo** [HeardLibrary/digital-scholarship#2]
Hi,
I successfully used your 2 Python scripts to populate my Wikibase instance, thank you!
It took me a while, but I also managed to modify the data type to have 'strings' instead of 'items' in the script _load_csv.py_. 
By the way, I would recommend suggesting in your tutorial [https://heardlibrary.github.io/digital-scholarship/host/wikidata/bot/] that the line 95 can be replaced by` 'value':'"' + stringValue + '"'` OR by `'"' + objectQNumber + '"'` (maybe not the best practice but it works) if we want to add the content of the csv column below _'P...'_ instead of one specific string literal. (And of course, examples for other types would help ;))

I was wondering to what extent the script can be adapted to deal with various types/values (url, string, items, external ID, etc.) contained in the same csv file? 
For the moment I can't see how to do it, but your script is the most crystal clear I've tested so far (compare with Piwikibot or WikidataIntegrator, for example), so I'd love to continue working with...
I hope you can help on this point and thx again for sharing (and commenting!) your code @baskaufs.

**Comment by baskaufs** [comment]
Hi @anchardo,
Thanks for your comment and I'm glad the script worked for you.  So am I understanding that you can just use the Q identifier for an item in place of a string value and the API will interpret it as an item?  

You are correct that the script as it currently stands is pretty limited.  I'm hoping to work on this more after August and one of the critical things would be to allow the spreadsheet to contain all of the different kinds of values that are allowed.  It would also be good to not hardcode the properties in the column headers.  One could use a sort of mapping file (similar to what I did [here](https://github.com/baskaufs/guid-o-matic/blob/master/buildings-column-mappings.csv), but maybe by putting the mappings into JSON instead of another CSV table.

**Comment by anchardo** [comment]
Thank for your answer @baskaufs.
Sorry it wasn't so clear.
In order to provide a string value for a property, you suggest in your tutorial to replace (line 95) :
`'value':'{"entity-type":"item","numeric-id":' + strippedQNumber+ '}'`
by 
`value':'"' + stringValue + '"`.
As I'm not using Python scripts often, it wasn't clear for me how to get a new variable for stringValue linked to the values stored in the csv file, rather than having 'stringValue' being literally used in Wikibase as value. (For example: line 95 : `'value':'"' + example1 + '"'` and 'example1' becoming the value in the Wikibase for every new statement).
By deduction, I understood that 'strippedQNumber' (on line 95) was based on 'objectQnumber' (lines 84-85) and that 'objectQnumber' would be the variable allowing me (line 95) to reuse what's inside the csv file as property values. (As I said I'm aware it's not a good practice and quite confusing to use the name 'objectQnumber' for a 'stringValue' but I wasn't sure if I could juste rename it without troubles). 
As all the other elements of the tutorial are very detailed/written in a step-by-step spirit, I think that may help users to clarify this 'stringValue' usage, as it very likely to have them in a Wikibase for specific properties.

Thank for letting me know about what you plan to do next. I'm 'following' your GitHub repo and I'll be thrilled if you succeed in implementing that!

**Comment by baskaufs** [comment]
Hi @anchardo I forgot all about this issue until I noticed it right now. I think that since we last discussed this, I've made a lot of the improvements to the API-writing script that you suggested. In particular, it handles all of the major data types except for Commons images (which I have not worked on yet). I've recently written some blog posts about it starting with [this one](http://baskauf.blogspot.com/2021/03/writing-your-own-data-to-wikidata-using.html). The posts all talk about Wikidata, but for the most part, they should all work with any Wikibase instance if you just change the API URL near the start of the script. 

The one type of thing that might not work is the feedback between the SPARQL endpoint and the script. I think that if the SPARQL endpoint is operating in the Wikibase instance, that should work, too, as long as you change the base URL of the endpoint from the Wikidata Query Service to the Wikibase SPARQL endpoint. If you are still working on this, I'd love to hear how it works with a generic Wikibase. That's something I've wanted to try, but haven't gotten around to yet.

Steve
